Add multi-fielded versions of the isNull and isNotNull functions to m…

jjonesy16 opened the pull request:
Ticket Details:  "The EvaluationPhaseFilterFunctions class has methods for single fielded isNull and isNotNull functions. When a multi-fielded isNull or isNotNull is in a query it is evaluated using the single fielded method and produces an incorrect evaluation."

- Added method overloads for EvaluationPhaseFilterFunctions IsNull and isNotNull that will except an Iterable<?> parameter and performed original method logic for each object found in new parameter.
- Added unit tests to EvaluationPhaseFilterFunctionsTest IsNotNullTests class.
- Added unit tests to EvaluationPhaseFilterFunctionsTest IsNullTests class.

jjonesy16 committed:
Add multi-fielded versions of the isNull and isNotNull functions to meet the requirements of Issue# 1604.

Ticket Details:  "The EvaluationPhaseFilterFunctions class has methods for single fielded isNull and isNotNull functions. When a multi-fielded isNull or isNotNull is in a query it is evaluated using the single fielded method and produces an incorrect evaluation."

jjonesy16 committed:
Fixed formatting w/formatter...

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:
```diff
@@ -145,6 +146,29 @@ public static FunctionalSet<ValueTuple> isNotNull(Object fieldValue) {
+     *            2nd field value to evaluate.
+     * @return the {@link FunctionalSet} of hit terms found
+     */
+    public static FunctionalSet<ValueTuple> isNotNull(Object fieldValue1, Object fieldValue2) {
```
This method signature should look like `isNotNull(Iterable<?> iterable)`. 

For an example, see the getSizeOf methods
- `getSizeOf(Object fieldValue)` (single fielded)
- `getSizeOf(Iterable<?> iterable)` (multi fielded)

jjonesy16 replied:
Hey Drew, thank you so much for the extra guidance!  This ticket makes a lot more sense now.  

I have one quick question...I'm struggling with the return type for the isNotNull methods.  Does the return type need to be of type, "FunctionalSet<ValueTuple>"?  To me, it seems that a True/False would be a better response from the isNotNull methods, but I also do not have a great understanding of how query side of the house either...  Any direction on this return type would be much appreciated.  The isNull is just a boolean, so that one makes good sense to me.

Thanks Buddy!

apmoriarty replied:
I think for the sake of consistency, the iterable version of isNotNull method will have to return a functional set just like it's single-fielded counterpart.

jjonesy16 replied:
Would a List<FunctionalSet> for the return type work?  That way I could put the FuctionalSet returned by the original method in a List for each item in the Iterable parameter passed in.  Let me know when you can, thanks!

jjonesy16 replied:
Modified code to utilize more appropriate method signature with multi-fielded filter parameters for the isNull and isNotNull overloaded methods.  Updated method signatures now utilize "Iterable<?> iterable" parameters.  Updated unit tests to test new method signatures and logic.

jjonesy16 replied:
Please review as time allows.  Thanks!

matthpeterson commented on warehouse/query-core/src/test/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctionsTest.java:
```diff
@@ -476,46 +476,97 @@ private boolean resultForDefaultOperator(int count) {
+            givenFieldValue1(valueTuple);
+            givenFieldValue2("F00.1");
+            
+            assertThat(resultForMultipleFieldValues()).containsExactly(valueTuple);
```
I'm wondering about "F00.1" - should it have been in the resultForMultipleFieldValues?  I'm not very familiar with the use cases for these FilterFunctions so please forgive my ignorance.

matthpeterson replied:
Is it that value1 and value2's contents overlap?

jjonesy16 replied:
Significant updates to original branch logic.  Please re-review as time permits.  Code in this comment has been replaced.

matthpeterson commented on warehouse/query-core/src/test/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctionsTest.java:
```diff
@@ -476,46 +476,97 @@ private boolean resultForDefaultOperator(int count) {
+            
+            assertThat(resultForMultipleFieldValues()).containsExactlyInAnyOrder(valueTuple1, valueTuple2);
+        }
+        
```
I'm reviewing the tests before the code and am thinking it might be worth adding more tests.  Here are a few scenarios to consider:
1. givenFieldValue1 is a collection containing multiple entries and givenFieldValue2 is empty
2. reverse of scenario 1, where value1 is empty and value2 is a collection
3. givenFieldValue1 is a collection containing multiple entries and givenFieldValue2 is null
4. reverse of scenario 3
5. reverse of testEmptyCollectionWithMultipleFields, where value2 is a non-empty collection and value1 is empty
6. reverse of testNullValueWithMultipleFields with null for 2, value 1 is a non-empty string
7. both value1 and value 2 are null
8. both value1 and value 2 are empty

jjonesy16 replied:
There have been significant changes from the initial MR, including the unit tests.  Can you please re-review when you have a chance.  Thanks!

matthpeterson commented on warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:
```diff
@@ -1,6 +1,7 @@
 package datawave.query.jexl.functions;
 
 import com.google.common.base.CharMatcher;
+import com.google.common.base.Function;
```
This might be an unused import.  I don't see Function referenced.

jjonesy16 replied:
Un-used imports removed.  Thanks!

matthpeterson commented on warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:
```diff
@@ -145,6 +146,29 @@ public static FunctionalSet<ValueTuple> isNotNull(Object fieldValue) {
+    }
+    
+    /**
+     * Returns a {@link FunctionalSet} of hit terms found for {@code fieldValue1}. If {@code fieldValue1} is a singular value tuple, a singleton
```
The first sentence may need to be adjusted unless I'm misunderstanding the intended functionality.  Can this method's comment be made to reference the single param method's comments, i.e. "behaves similiarly to isNotNull(Object) except on multiple parameters.  When evaluating multiple parameters, every parameter must meet the criteria or an emptySet will be returned."  I'm not sure that's accurately describing the intention.

jjonesy16 replied:
Modified method comments to have a better description.  Thank you for the suggestion!

matthpeterson commented on warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:
```diff
@@ -145,6 +146,29 @@ public static FunctionalSet<ValueTuple> isNotNull(Object fieldValue) {
+        FunctionalSet<ValueTuple> matches = isNotNull(fieldValue1);
+        
+        if (matches != FunctionalSet.EMPTY_SET && !((FunctionalSet) isNotNull(fieldValue2)).isEmpty()) {
+            return matches;
```
If it's a reasonable goal to include matching terms for fieldValue2, the results of isNotNull on each of the parameters could be merged together.

jjonesy16 replied:
There have been significant changes from the initial MR, including the unit tests.  Can you please re-review when you have a chance.  Thanks!

matthpeterson commented on warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:
```diff
@@ -145,6 +146,29 @@ public static FunctionalSet<ValueTuple> isNotNull(Object fieldValue) {
+    public static FunctionalSet<ValueTuple> isNotNull(Object fieldValue1, Object fieldValue2) {
+        FunctionalSet<ValueTuple> matches = isNotNull(fieldValue1);
+        
+        if (matches != FunctionalSet.EMPTY_SET && !((FunctionalSet) isNotNull(fieldValue2)).isEmpty()) {
```
It might be prudent to use ! FunctionalSet.EMPTY_SET.equals(matches) instead of !=.

jjonesy16 replied:
There have been significant changes from the initial MR, including the unit tests. Can you please re-review when you have a chance. Thanks!

matthpeterson commented on warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:
```diff
@@ -159,6 +183,25 @@ public static boolean isNull(Object fieldValue) {
+     * @return true if {@code fieldValue1} or {@code fieldValue2} is a null {@link Object} or an empty {@link Collection}, or false otherwise
+     */
+    public static boolean isNull(Object fieldValue1, Object fieldValue2) {
+        if (isNull(fieldValue1)) {
```
This is just a stylistic thought which doesn't require any change.  This method could be modified to be:
```
return isNull(fieldValue1) || isNull(fieldValue2);
```

jjonesy16 replied:
There have been significant changes from the initial MR, including the unit tests. Can you please re-review when you have a chance. Thanks!

jjonesy16 committed:
Modified isNotNull overloaded method to utilize parameter type Iterable<?> for use cases with multiple input parameters.

Updated unit tests to reflect isNotNull method changes.

Minor changes to isNull overloaded method and unit tests per peer review feedback.

jjonesy16 committed:
Modified IsNotNull overloaded method logic to have the same return type as original method.  New logic will return 1st FunctionalSet<ValueTuple> found within the Iterable<?> input parameter as long as no null values are encountered in the Iterable.  If a null value is found while processing Iterable<?> parameter, an empty FunctionalSet is returned.

jjonesy16 committed:
#1715 & #1604 - Implemented previously written unit tests for #1604 in DatawaveInterpreterTest.java with new logic in EvaluationPhaseFilterFunctions.java isNull and isNotNull overloaded methods.

apmoriarty commented:
I was wrong when I wrote the original ticket. This is not as simple as adding iterable versions of the isNull/isNotNull functions. There is additional work required in the DatawaveInterpreter to detect these functions and then pass in the correct hits per field. Let's discuss later today if you have time.

avgAGB commented on warehouse/query-core/src/test/java/datawave/query/jexl/DatawaveInterpreterTest.java:
```diff
@@ -297,32 +296,27 @@ public void testFilterFunctionIsNull() {
         
         // multi field, all present
-        String query = "FOO == 'bar' && filter:isNull(FOO || FOO)";
+        String query = "FOO == 'bar' && (filter:isNull(FOO) || filter:isNull(FOO))";
```
To clarify here, will the checks eventually be returned to the original logic once the ticket here is merged, or will it be dependent on #1604  still?

apmoriarty commented:
Close this PR per discussion comments
